stan/math/rev/functor/cvodes_integrator.hpp

```diff
@@ -75,116 +183,174 @@ class cvodes_integrator {
     }
 
-    const size_t coupled_size = cvodes_data.coupled_ode_.size();
-
-    std::vector<std::vector<return_type_t<T_initial, T_param, T_t0, T_ts>>> y;
-    coupled_ode_observer<F, T_initial, T_param, T_t0, T_ts> observer(
-        f, y0, theta, t0, ts, x, x_int, msgs, y);
-
     try {
-      check_flag_sundials(CVodeInit(cvodes_mem, &ode_data::cv_rhs, t0_dbl,
-                                    cvodes_data.nv_state_),
+      check_flag_sundials(CVodeInit(cvodes_mem, &cvodes_integrator::cv_rhs,
+                                    value_of(t0_), nv_state_),
                           "CVodeInit");
 
       // Assign pointer to this as user data
       check_flag_sundials(
-          CVodeSetUserData(cvodes_mem, reinterpret_cast<void*>(&cvodes_data)),
+          CVodeSetUserData(cvodes_mem, reinterpret_cast<void*>(this)),
           "CVodeSetUserData");
 
-      cvodes_set_options(cvodes_mem, relative_tolerance, absolute_tolerance,
-                         max_num_steps);
+      cvodes_set_options(cvodes_mem, relative_tolerance_, absolute_tolerance_,
+                         max_num_steps_);
 
       // for the stiff solvers we need to reserve additional memory
       // and provide a Jacobian function call. new API since 3.0.0:
       // create matrix object and linear solver object; resource
       // (de-)allocation is handled in the cvodes_ode_data
+      check_flag_sundials(CVodeSetLinearSolver(cvodes_mem, LS_, A_),
+                          "CVodeSetLinearSolver");
       check_flag_sundials(
-          CVodeSetLinearSolver(cvodes_mem, cvodes_data.LS_, cvodes_data.A_),
-          "CVodeSetLinearSolver");
-      check_flag_sundials(
-          CVodeSetJacFn(cvodes_mem, &ode_data::cv_jacobian_states),
+          CVodeSetJacFn(cvodes_mem, &cvodes_integrator::cv_jacobian_states),
           "CVodeSetJacFn");
 
       // initialize forward sensitivity system of CVODES as needed
-      if (S > 0) {
+      if (y0_vars_ + args_vars_ > 0) {
         check_flag_sundials(
-            CVodeSensInit(cvodes_mem, static_cast<int>(S), CV_STAGGERED,
-                          &ode_data::cv_rhs_sens, cvodes_data.nv_state_sens_),
+            CVodeSensInit(cvodes_mem, static_cast<int>(y0_vars_ + args_vars_),
+                          CV_STAGGERED, &cvodes_integrator::cv_rhs_sens,
+                          nv_state_sens_),
             "CVodeSensInit");
 
-        check_flag_sundials(CVodeSensEEtolerances(cvodes_mem),
-                            "CVodeSensEEtolerances");
+        if (include_sensitivities_in_errors_) {
+          check_flag_sundials(CVodeSetSensErrCon(cvodes_mem, SUNTRUE),
+                              "CVodeSetSensErrCon");
+        } else {
+          check_flag_sundials(CVodeSetSensErrCon(cvodes_mem, SUNFALSE),
+                              "CVodeSetSensErrCon");
+        }
+
+        if (atols.size() > 0) {
+          check_positive_finite("cvodes_integrator",
```

Comment:
for the sake of time at this stage, we should drop this functionality for now. We can always add this later again.

Moreover, I realised that users can *implicitly* add this feature by arbitrary scaling of the ode states. That's not quite the same, but it's getting close to this, I think.